client/main.js

```diff
@@ -29,73 +29,94 @@ if ( global.wcConnectData ) {
+		case 'wc-connect-shipping-settings':
+			return ShippingSettings;
+		case 'wc-connect-admin-test-print':
+			return PrintTestLabel;
+		case 'wc-connect-stripe-connect-account':
+			return StripeConnectAccount;
+		default:
+			return null;
+	}
+};
+
+const createdStores = {};
+
 Array.from( document.getElementsByClassName( 'wcc-root' ) ).forEach( ( container ) => {
 	const args = container.dataset.args && JSON.parse( container.dataset.args ) || {};
 	delete container.dataset.args;
-
-	const RouteClass = getRouteClass( container.classList );
+	const routeClassName = getRouteClassName( container.classList );
+	
+	const RouteClass = getRouteClass( routeClassName );
 	if ( ! RouteClass ) {
 		return;
 	}
 	const Route = RouteClass( args );
 
-	const persistedStateKey = Route.getStateKey();
-	const persistedState = storageUtils.getWithExpiry( persistedStateKey );
-	storageUtils.remove( persistedStateKey );
-	const serverState = Route.getInitialState();
-	const initialState = { ...serverState, ...persistedState };
+	if( typeof createdStores[ routeClassName ] === 'undefined' ) {
+		const persistedStateKey = Route.getStateKey();
+		const persistedState = storageUtils.getWithExpiry( persistedStateKey );
+		storageUtils.remove( persistedStateKey );
+		const serverState = Route.getInitialState();
+		const initialState = { ...serverState, ...persistedState };
 
-	const middlewares = [
-		thunk.withExtraArgument( args ),
-		wpcomApiMiddleware,
-		localApiMiddleware,
-	];
+		const middlewares = [
+			thunk.withExtraArgument( args ),
+			wpcomApiMiddleware,
+			localApiMiddleware,
+		];
 
-	if ( Route.getMiddlewares ) {
-		middlewares.push.apply( middlewares, Route.getMiddlewares() );
-	}
+		if ( Route.getMiddlewares ) {
+			middlewares.push.apply( middlewares, Route.getMiddlewares() );
+		}
 
-	const enhancers = [
-		applyMiddleware( ...middlewares ),
-		window.devToolsExtension && window.devToolsExtension(),
-	].filter( Boolean );
+		const enhancers = [
+			applyMiddleware( ...middlewares ),
+			window.devToolsExtension && window.devToolsExtension(),
+		].filter( Boolean );
 
-	const store = compose( ...enhancers )( createStore )( Route.getReducer(), initialState );
+		const store = compose( ...enhancers )( createStore )( Route.getReducer(), initialState );
 
-	if ( Route.getInitialActions ) {
-		Route.getInitialActions().forEach( store.dispatch );
-	}
+		if ( Route.getInitialActions ) {
+			Route.getInitialActions().forEach( store.dispatch );
+		}
 
-	window.addEventListener( 'beforeunload', () => {
-		const state = store.getState();
+		window.addEventListener( 'beforeunload', () => {
+			const state = store.getState();
 
-		if ( window.persistState ) {
-			storageUtils.setWithExpiry( persistedStateKey, Route.getStateForPersisting( state ) );
-		}
-	} );
+			if ( window.persistState ) {
+				storageUtils.setWithExpiry( persistedStateKey, Route.getStateForPersisting( state ) );
+			}
+		} );
+
+		createdStores[ routeClassName ] = store;
```

Comment:
Curious why we want to use multiple stores instead of just 1 store?

Reply:
Yep! Agree with you. @budzanowski said he'll bring it to our next roundtable for discussion.

Reply:
Oh I see! Thanks!

Reply:
@harriswong  Yeh, this architecture is weird. This is what we have inherited - multiple stores per different components. This makes working with the whole app very counterintuitive. I have merged the store for components of the same type. Which looks like a step 1. Step 2 will be to have everything under the same Redux store. Right now different components can use different middleware - maybe this is something that is actually required - but I doubt that it is really useful. After my changes various QueryComponents located in different parts of the application started to communicate ( through a shared store ) and stopped duplicating API calls. The label purchase code is a bit more snappy now :)